jupyterlab_gcedetails/src/service/notebook_instance_service_layer.ts

```diff
@@ -0,0 +1,453 @@
+          }
+        } catch (err) {
+          if (++attempt === POLL_RETRIES) {
+            console.error(
+              `Unable to retrieve Operation status from ${path} after ${attempt} attempts`
+            );
+            clearInterval(interval);
+            reject(err);
+          }
+        }
+      }, POLL_INTERVAL);
+    });
+  }
+
+  /**
+   * Returns the projectId from the object or tries to retrieve it from the
+   * server if not set.
+   */
+  private async _getProject(): Promise<string> {
+    if (this.projectIdPromise) {
+      return this.projectIdPromise;
+    }
+    try {
+      const { project } = await this._getMetadata();
+      this.projectIdPromise = Promise.resolve(project);
+      return this.projectIdPromise;
+    } catch (err) {
+      console.error('Unable to obtain GCP Project', err);
+      throw err;
+    }
+  }
+
+  /**
+   * Returns the instanceName from the object or tries to retrieve it from the
+   * server if not set.
+   */
+  private async _getInstance(): Promise<string> {
+    if (this.instanceNamePromise) {
+      return this.instanceNamePromise;
+    }
+    try {
+      const { name } = await this._getMetadata();
+      this.instanceNamePromise = Promise.resolve(name);
+      return this.instanceNamePromise;
+    } catch (err) {
+      console.error('Unable to obtain GCP Instance Name', err);
+      throw err;
+    }
+  }
+
+  /**
+   * Returns the locationId from the object or tries to retrieve it from the
+   * server if not set.
+   */
+  private async _getLocation(): Promise<string> {
+    if (this.locationIdPromise) {
+      return this.locationIdPromise;
+    }
+    try {
+      const { zone } = await this._getMetadata();
+      // extract the location from the zone resource name
+      const locationId = zone.split('/')[3];
+      this.locationIdPromise = Promise.resolve(locationId);
+      return this.locationIdPromise;
+    } catch (err) {
+      console.error('Unable to obtain GCP Zone', err);
+      throw err;
+    }
+  }
+
+  /**
+   * Retrieves the VM Metadata from the server
+   */
+  private _getMetadata(): Promise<InstanceMetadata> {
+    if (this.metadataPromise) {
+      return this.metadataPromise;
+    }
+    this.metadataPromise = getMetadata().catch(err => {
+      console.error('Unable to obtain VM Metadata');
+      throw err;
+    });
+    return this.metadataPromise;
+  }
+
+  /**
+   * Stops the notebook instance.
+   */
+  async stop(): Promise<Instance> {
+    try {
+      const [projectId, instanceName, locationId] = await Promise.all([
+        this.projectId,
+        this.instanceName,
+        this.locationId,
+      ]);
+      const name = `projects/${projectId}/locations/${locationId}/instances/${instanceName}`;
+      const pendingOperation = await this._transportService.submit<Operation>({
+        path: `${NOTEBOOKS_API_PATH}/${name}:stop`,
+        method: POST,
+        headers: { Authorization: `Bearer ${this._authToken}` },
```

Comment:
This may be problematic. If you're using the `ServerProxyTransportService`, it's going to automatically add its own `Authorization` header based on the machine's credential. 

The [ClientTransportService](https://github.com/GoogleCloudPlatform/jupyter-extensions/blob/2b51a62b78c549bba72668e181acdccf0f217acc/shared/src/service/transport.ts#L98) is a better choice, but it forces the user to go through the OAuth flow, which we don't want. We may need to modify that class to allow an OAuth token to be passed in, which would bypass the initialization flow.

Reply:
Ah I noticed that and was going to ask you how this the authorization plays out in our use case, since while testing it did seem to work. We can talk more about this during stand-up but it's a good call to use ClientTransportService as you suggested - I'll add that!